core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/zookeeper/install-upgrade.yml

```
@@ -0,0 +1,80 @@
+---
+
+# This task can be reused from zookeeper role. Rememeber about it in near future. We should refactor zookeeper role and split some code into smaller pieces.
+- name: Download Zookeeper binaries
+  include_role:
+    name: download
+    tasks_from: download_file
+  vars:
+    file_name: "{{ zookeeper_defaults.zookeeper_bin_filename }}"
+
+- name: Create Zookeeper directories
+  become: yes
+  file:
+    path: "/opt/zookeeper-{{ zookeeper_defaults.zookeeper_version }}"
+    recurse: yes
+    owner: "{{ zookeeper_defaults.zookeeper_user }}"
+    group: "{{ zookeeper_defaults.zookeeper_group }}"
+    mode: u=rwx,g=rx,o=rx
+    state: directory
+
+- name: Uncompress Zookeeper installation file
+  unarchive:
+    remote_src: yes
+    src: "{{ download_directory }}/{{ zookeeper_defaults.zookeeper_bin_filename }}"
+    dest: /opt/zookeeper-{{ zookeeper_defaults.zookeeper_version }}
+    creates: "/opt/zookeeper-{{ zookeeper_defaults.zookeeper_version }}/bin"
+    extra_opts: [--strip-components=1]
+    mode: u=rwx,g=rx,o=rx
+    owner: "{{ zookeeper_defaults.zookeeper_user }}"
+    group: "{{ zookeeper_defaults.zookeeper_group }}"
+
+- name: Copy configuration from previous version
+  copy:
+    remote_src: yes
+    src: /opt/zookeeper/conf/
+    dest: /opt/zookeeper-{{ zookeeper_defaults.zookeeper_version }}/conf
+    mode: preserve
+
+- name: Link /opt/zookeeper to recently installed version
+  file:
+    dest: /opt/zookeeper
+    state: link
+    src: /opt/zookeeper-{{ zookeeper_defaults.zookeeper_version }}
+    force: yes
+  register: new_symbolic_link
+
+- name: Remove previous version binaries
+  file:
+    path: /opt/zookeeper-{{ before_upgrade_zookeeper_version }}
+    state: absent
+  when: new_symbolic_link.changed
```

```suggestion
  when: before_upgrade_zookeeper_version != zookeeper_defaults.zookeeper_version
```
This is clean up task so I would move it somewhere after starting ZK service. I think it's better to remove old version only after the new one is working 😉